design-proposals/klip-4-custom-types.md

```diff
@@ -0,0 +1,154 @@
+  # KLIP-4: Custom Type Registry
+
+**Author**: agavra | 
+**Release Target**: 5.4 | 
+**Status**: In Discussion | 
+**Discussion**: link
+
+**tl;dr:** *Introduce a feature that makes custom types easier to work with in KSQL by aliasing
+complex type declarations.*
+
+## Motivation and background
+
+Kafka often contains very complex nested data structures and it is common to see schemas nested with
+other schemas. It is desirable to be able to define this custom schema in KSQL once and reference it
+in all future statements. For example, imagine the following schema declarations:
+
+```sql
+REGISTER TYPE ADDRESS AS STRUCT<number INTEGER, street VARCHAR, city VARCHAR>;
+REGISTER TYPE PERSON AS STRUCT<firstname VARCHAR, lastname VARCHAR, address ADDRESS>;
+REGISTER TYPE COMPANY AS STRUCT<name VARCHAR, headquarters ADDRESS>;
+```
+
+With this feature organizations can leverage existing schemas in their KSQL queries and share 
+data with other parts of their organizations.
+
+## Scope
+
+* Syntax to register and unregister custom type aliases
+* Ability to include custom types at compile-time (e.g. in the JAR) for use in UDF declarations
+
+## Value/Return
+
+Real data is complex, nested and referential. This feature will help us dramatically improve the
+user interaction with complicated schemas. 
+
+## Public APIS
+
+### REGISTER TYPE
+
+```sql
+REGISTER TYPE <type_name> AS <type>;
+```
+
+The `REGISTER TYPE` syntax will allow KSQL users to register a type alias directly in SQL (either 
+interactive or headless modes). Any types registered using this command can be leveraged in any
+future statement. 
+
+Any attempts to register the same type twice without a corresponding `UNREGISTER TYPE` will fail.
+
+### UNREGISTER TYPE
+
+```sql
+UNREGISTER TYPE <type_name>;
+```
+
+The `UNREGISTER TYPE` syntax will allow KSQL users to remove a type alias from KSQL. This statement
+will fail if the type is being used in any active query or UDF.
+
+### Extension Directory
+
+The `ksql.extension.dir` will now also recognize `.sql` files that contain only `REGISTER TYPE`
+commands. These commands will all be run before compiling any UDFs so these custom type declarations
+can be used in the `@Udf` annotations.
+
+## Design
+
+There will be a rewriter phase during parsing that will replace any aliased type with the full
```

Comment:
This rewrite phase will be in the engine, right? The parser should be kept clean?

I think in time the parser should not even know about the metastore.

Reply:
Yes - that would make sense. I didn't really intend to design the code structure with this KLIP.

Reply:
The more I look at it, the less likely it is that we will be able to have the rewrite stage independent of the parser. The parser needs to know which SQL type each type is, and without information on the registered types it won't be able to do that. So without a major refactor (moving type system outside of the parser) we need to pass this in.